### PR TITLE
Support for trim path effects in Flutter

### PR DESCRIPTION
**Pull request by luigi-rosso**
We're adding support for trimmed paths in Flare and Flutter. This allows for really interesting effects like rain drops, fireworks, and a lot of other cases where being able to animate the visible portion of the stroke is useful.

Here is an example of a square with a trimmed stroke in Flare:
screen shot 2019-01-15 at 11 19 20 am

Here's a terribly compressed gif  of it running in Flutter using an engine with the modifications in this pull request.
trimpath

### CODE REVIEW

**Comment by luigi-rosso** [comment]
Here's a better example in Flare of how powerful this feature can be, almost all the animation here is done by keying trim path properties:
https://twitter.com/guidorosso/status/1085353625671483392

File: lib/ui/painting.dart [hunk 1]
```diff
@@ -2079,6 +2079,16 @@ class Path extends NativeFieldWrapperClass2 {
   }
   Path _transform(Float64List matrix4) native 'Path_transform';
 
+  /// Cuts the path into a subset of itself leaving only the region between
+  /// `startT` and `stopT` fraction values. Set `isComplement` to true to
+  /// invert the result by cutting out the region from `startT` to `stopT`.
+  ///
+  /// Returns true when the cut is succesfully made, false if not.
+  bool trim(double startT, double stopT, bool isComplement) {
```
Comment (chinmaygarde) [inline lib/ui/painting.dart]: Please document the valid ranges and also clamp before making the native call.

Comment (chinmaygarde) [inline lib/ui/painting.dart]: Can isComplement just be isInverted to match Skia terminology?

Comment (luigi-rosso) [reply lib/ui/painting.dart]: Yes, that is definitely more clear.

Comment (luigi-rosso) [reply lib/ui/painting.dart]: Sounds good, I changed it to the same range provided in the Skia docs. I originally didn't clamp them as Skia does this internally, but I did make the change now!

**Comment by dnfield** [comment]
When this has come up in the past, it's been suggested we just do this on the Dart side, e.g. https://github.com/dnfield/flutter_path_drawing/blob/master/lib/src/trim_path.dart

Do we have any metrics on how much this will increase binary size?  I would rather see Skia effects implemented that we can't easily do in Dart....

**Comment by dnfield** [comment]
(in other words, effects that can't be implemented by simply using `PathMetrics`)

**Comment by dnfield** [comment]
Dashing is also implemented in the linked package, although I haven't done any significant tests of it.  There's signfiicant discussion on that at https://github.com/flutter/flutter/issues/4858

**Comment by luigi-rosso** [comment]
I did notice you had implemented it, that's what sparked my curiosity as we are designing that functionality into Flare next week. Thanks for sharing your code. 

I'm going to close this PR. The issue you linked makes an excellent case for keeping code like this (and dashing) in dart. Thanks to you both for the feedback.

**Comment by dnfield** [comment]
No problem!  Thanks for the effort, and I like what you guys are doing with Flare a lot!

Feel free to let me know if you find issues with the flutter_path_drawing lib.  I'm using it for flutter_svg and a little bit in the lottie-flutter implementation (although that one needs a lot of love).

**Comment by luigi-rosso** [comment]
Awesome, thank you! Great work on all those libraries! 

I ended up rolling something more specific to our needs for the trim path as we needed it to allow for trimming from variable start points (and @neurowave wanted it to loop around if you go past the end). Does lottie only trim from the start/end of a path?

I implemented something very similar to what Skia does internally: https://github.com/2d-inc/Flare-Flutter/blob/master/flare_flutter/lib/trim_path.dart

One feature I'd like to add is the option (in Flare) to trim sub-paths/multiple contours sequentially or in sync. Right now we only do them sequentially which requires measuring twice while trimming :( We can definitely improve that by detecting when it's a single contour, however.

I'll be taking a good look at your dashing implementation! Seems pretty lean and mean at first glance.

**Comment by luigi-rosso** [comment]
Hideous example: trimming

**Comment by neurowave** [comment]
Yes! The offset feature (see @luigi-rosso’s bottom row of keys) is a lot more useful if it loops! I can create some specific examples to illustrate this, if it helps.

**Comment by dnfield** [comment]
Atually I think the lottie implementation is different.  I'd be happy to extend the library, but there's nothing super magical about any of this stuff, and the trim path in there is more of a demo.  I'd be happy to enhance it if it helps, but it's also fine if you guys have something working and don't want an extra dependency.

**Pull request closed without merging**